relayer/app/cursors/watcher.go

```diff
@@ -0,0 +1,123 @@
+package cursors
+
+import (
+	"context"
+	"slices"
+
+	"github.com/omni-network/omni/lib/errors"
+	"github.com/omni-network/omni/lib/xchain"
+)
+
+// todo mind the query limit on db
+
+type Watcher struct {
+	cursors  CursorTable
+	provider xchain.Provider
+}
+
+func NewWatcher(cursors CursorTable, provider xchain.Provider) (*Watcher, error) {
+	return &Watcher{
+		cursors,
+		provider,
+	}, nil
+}
+
+func (w *Watcher) getStreamCursors(ctx context.Context, stream xchain.StreamID) ([]*Cursor, error) {
+	iterator, err := w.cursors.List(
+		ctx,
+		CursorPrimaryKey{}.WithSrcChainIdConfLevelDstChainId(
+			stream.SourceChainID,
+			uint32(stream.ConfLevel()),
+			stream.DestChainID,
+		))
+
+	if err != nil {
+		return nil, errors.Wrap(err, "list cursors")
+	}
+	defer iterator.Close()
+
+	var cursorsDesc []*Cursor
+	for iterator.Next() {
+		cursor, err := iterator.Value()
+		if err != nil {
+			return nil, errors.Wrap(err, "cursor value")
+		}
+		cursorsDesc = append(cursorsDesc, cursor)
+	}
+
+	return cursorsDesc, nil
+}
+
+// ConfirmStream checks all cursors for the provided stream and confirms them according to the rules:
+//   - if a cursor is not empty fetch finalized cursor from the network and confirm only if finalized
+//     message offset is higher to the cursor last message offset
+//   - if a cursor is empty, confirm it if the previous cursor was confirmed
+func (w *Watcher) ConfirmStream(ctx context.Context, stream xchain.StreamID) error {
+	cursors, err := w.getStreamCursors(ctx, stream)
+	if err != nil {
+		return err
+	}
+
```

Comment:
why not something like:
```
for i, c := range cursors {
  if c.Confirmed {
    continue
  }
  if !c.Empty() {
    finalized, err := isFinalized(ctx, c)
    if !finalized {
      return nil // Reached head
    }
  }
  
  setConfirmed(ctx, c)
}

```

Reply:
because you would have to fetch cursors from the network for each non-empty cursor even though you know what the latest finalized cursor is

Reply:
unless you cache the network request in isFinalized but then it's basically what I wrote. And if you suggest just extracting into more functions, that can be changed, I plan to go over that in the end again, but extracting can be improved

Reply:
at this point attestations are few and far between, not sure caching is required. I'd just stick to simple code rather?

Reply:
be careful of #prematureoptimisations

Reply:
sure we can always fetch, but in any case the logic above won't work because we can't just set to confirmed all, what if the cursors are all empty, you cant just confirm them, until you have one non-empty confirmed, above suggestion would just confirm empty ones.